### PR TITLE
Added generation of indices and further analytical functions

### PR DESCRIPTION
**Pull request by nmueller18**
The new branch contains the following:
- new indices for use in other functions
- check of representativity of data for younger age groups
- correction of life tables
- reproduction rates (differs in some instances from Hinz 2011)
- Masculinity index and maternal mortality

I think there is a lot to explore, and some things might even be useful :-)

### CODE REVIEW

File: R/further_analytical_functions.R [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+#'
+#' @export
+#'
+lt.correction <- function(life_table, agecor = TRUE, agecorfac = c(), option_spline = NULL) {
+
+  indx <- lt.indices(life_table)
+
+  # corrected life expectancy at birth after Bocquet-Appel and Masset
+  e0 <- 78.721 * log10(sqrt(1 / indx$juvenile_i$d5_14_d20plus)) - 3.384
+  e0_range <- paste0(round(e0 - 1.503,1), "--", round(e0 + 1.503,1))
```
Comment (nevrome) [inline R/further_analytical_functions.R]: Maybe it's better not to make the range a string value. When somebody wants to do anything else with this value beyond simply looking at it its's more convenient to be able to access it directly. 

Maybe something like this would be better:

|   | value | range_start | range_stop |
|---|---|---|---|
| e0   |   |   |   |
| 1q0 |   |   |   |
| 5q0  |   |   |   |

File: R/further_analytical_functions.R [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+  life_table_prep_corrected <- data.frame(cbind(a = life_table$a, Dx = life_table$Dx))
+  life_table_corr <- life.table(life_table_prep_corrected, agecor = agecor, agecorfac = agecorfac, option_spline = option_spline)
+  row_e0 <- c(round(e0, 1), e0_range)
+  row_q1_0 <- c(round(q1_0, 3), q1_0_range)
+  row_q5_0 <- c(round(q5_0, 3), q5_0_range)
+  e0_q5_0 <- data.frame(rbind(row_e0, row_q1_0, row_q5_0))
+  colnames(e0_q5_0) <- c("value", "range")
+  rownames(e0_q5_0) <- c("e0", "1q0", "5q0")
+  return(list(indices = e0_q5_0, life_table_corr = life_table_corr))
+  }
```
Comment (nevrome) [inline R/further_analytical_functions.R:118]: Some more comments in the code would be nice.

File: R/further_analytical_functions.R [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+
+#' Calculate masculinity index and maternal mortality
+#' from a mortAAR life table list
+#'
+#' The proportional relation between adult males and females (=
+#' Masculinity index) is interesting for a number of reasons: (1)
+#' it can point to basic problems in the datasets in that, say,
+#' one sex is grossly over- or underrepresented. (2) it may hint
+#' towards cultural reaons like sex-specific mobility.\cr
```
Comment (nevrome) [inline R/further_analytical_functions.R]: spelling: reasons

File: R/further_analytical_functions.R [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+#' Maternal mortality is a basic indicator for the health system of a
+#' given population. Recently, McFadden and Oxenham (2019) have
```
Comment (nevrome) [inline R/further_analytical_functions.R]: Full citation should be added to the help file. Maybe with the solution we used for life.table. (Rdpack?)

File: R/further_analytical_functions.R [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+#' an MI < 100 does not necesssarily speak for an unbalanced MI in life.\cr
+#' Maternal mortality is calculated according to the formula
+#' provided by McFadden and Oxenham (2019). McFadden and Oxenham show
+#' that with modern data a very high correlation is achieved by only
+#' comparing the absolute numbers of the age group 20 to 24. This has the
+#' additional advantage that for this age group anthropological aging
+#' methods are reasonable exact.
+#'
+#' @param life_table an object of class mortaar_life_table_list.
+#' @param females an object of class mortaar_life_table for females
```
Comment (nevrome) [inline R/further_analytical_functions.R:146]: This is just the group name as a character. Not the object. 

Maybe it's better to change the input: No life.table.list which contains the two life.tables + the relevant names, but just the two relevant life.tables.

File: R/further_analytical_functions.R [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+#' .$age == "7-8"    ~ "7-9", .$age == "9-10"   ~ "9-11",
+#' TRUE ~  .$age))
+#' muen <- muen %>% tidyr::separate(age, c("from", "to")) %>%
+#' transform(from = as.numeric(from), to = as.numeric(to))
+#' muen_prep <- muen %>% prep.life.table(group = "sex",
+#' agebeg = "from", ageend = "to", method = "Standard",
+#' agerange = "excluded")
+#' muen_result <- muen_prep %>% life.table()
+#' lt.sexrelation(muen_result,"female","male")
+#'
```
Comment (nevrome) [inline R/further_analytical_functions.R:173]: This example code is a bit overwhelming. Too specific and too much stuff that is not related to `lt.sexrelation`. Maybe it's better to create a new example dataset that can be used directly with this function. Without the tedious preparation.

This also applies for the other functions.

File: R/further_analytical_functions.R [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+  if (!inherits(life_table,"mortaar_life_table_list"))  {
+    paste0(
+      "The dataset does not seem to be a mortAAR life table list."
+    ) %>%
+      message
+  } else {
+
+    # Masculinity index (MI) for juvenile and older individuals: males * 100 / females
+
+    fem_age <- life_table[females][[1]][[2]] %>% cumsum
```
Comment (nevrome) [inline R/further_analytical_functions.R]: It's better to access values in the tables with names and not indices. When something changes, debugging with indices only is hell, because you have no idea what the number meant in the past.

File: R/further_analytical_functions.R [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+    m_mort <- life_table[males][[1]][[4]][which(life_table[males][[1]][[1]]=="20--24")]
+    maternal_mortality <- (333.33 * f_mort / m_mort) - 76.07
+    maternal_mortality_rate <- round(100000 / maternal_mortality)
+
+    cat(sep='', "Masculinity index: \n", masculinity_index,
+        "\n\nMaternal mortality:\n", round(maternal_mortality,1),
+        "\n\nOne of ", maternal_mortality_rate, " women is dying during",
+        " pregnancy or within the first 42 days after birth due to",
+        " complications.")
+  }
```
Comment (nevrome) [inline R/further_analytical_functions.R:211]: 1. Please avoid `cat` everywhere. You should use `message` for console messages.
2. No function should only print a message. The values should also be numerically accessible. `lt.sexrelation` should return a list with the two output values.

File: R/further_analytical_functions.R [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+#' For population studies it is of vital importance to estimate growth or
+#' decline of a population. For archaeological datasets this is rarely
+#' attempted, probably because the data quality seems to scanty.
+#' Nevertheless, the calculation of such measures seems worth the try,
+#' at least it should give an impression if the resulting values
+#' are unrealistic high or low.
+#'
+#' There are different approaches to calculate reproduction rates (e. g.
+#' Henneberg 1978). We largely follow the methodology by Hassan 1981.
+#' Typically, a Total fertility rate (TFR) of 6-8 is assumed for
```
Comment (nevrome) [inline R/further_analytical_functions.R]: Citation as mentioned above.

File: R/further_analytical_functions.R [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+
+  fertiliy_rate <- c(round(fertil_rate, 1), "Total fertility rate")
+  R_pot_fem <- c(round(R_pot_fem, 1), "Gross reproduction rate")
+  R_0 <- c(round(R_0, 2), description = "Net reproduction rate")
+  intr_grow <- c(round(intr_grow, 2), "Intrinsic growth rate (perc/y)")
+  Dt <- c(round(Dt, 1), "Doubling time in years")
+  result <- data.frame(rbind(fertiliy_rate, R_pot_fem, R_0, intr_grow, Dt))
+  colnames(result) <- c("value", "description")
+  rownames(result) <- c("TFR","GRR", "NRR","r", "Dt")
+  format(result, justify = "left")
```
Comment (nevrome) [inline R/further_analytical_functions.R]: Not necessary. `return(result)` is much better readable.

Do not care too much about how something might look like on somebodies console. You can't control that anyway. But that's just my opinion.

File: R/lifetable_indices.R [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+      "The dataset does not seem to be a mortAAR life table."
+    ) %>%
+      message
+  } else if (!inherits(life_table,"mortaar_life_table"))  {
+    paste0(
+      "Please select only one life table."
+    ) %>%
+      message
+  }
+}
```
Comment (nevrome) [inline R/lifetable_indices.R:17]: We already have `is.mortaar_life_table_list` and `is.mortaar_life_table`. Do we need this function?

File: R/lifetable_indices.R [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+#'   \item \bold{D0_14_D}:   proportion of individuals aged 0--14
+#'   according to McFadden and Oxenham 2017 if infants are represented
+#'   well.
+#'   \item \bold{mortality}:   list of mortalities 5q0, 5q10, 5q15, 45q15.
+#'   \item \bold{repro_indx}: vector of survival rates of individuals
+#'   aged 15--45.
+#'   \item \bold{e0}:   life expectancy at age 0.
+#'}
+#'
+#' @export
```
Comment (nevrome) [inline R/lifetable_indices.R:51]: Is it necessary to export this function? Seems to be a helper. Maybe it would be better to split this into many small functions.

File: R/representativity.R [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    crayon::red(paste0(representativity_sum, " of 4 criteria for representativity of subadults are NOT given. You shoud refrain from any corrections."))
+  } else if (representativity_sum == 1) {
+    crayon::red(paste0("One of 4 criteria for representativity of subadults is NOT given. You shoud be careful with any corrections."))
+  } else {
+    crayon::green(paste0("General representativity of subadults seems to be given. Further corrections might be applied."))
+  }
+
+  cat("Criteria for representativity after Weiss 1973:\n", weiss_i1, "\n", weiss_i2,
+      "\n\nCriteria for representativity after Bocquet-Appel and Masset:\n", child_i, "\n", juvenile_i,
+      "\n\nRepresentativity verdict:\n",representativity_verdict)
```
Comment (nevrome) [inline R/representativity.R]: As already mentioned above: That's not a good way to provide the output of a function.

Comment (nevrome) [reply R/representativity.R]: If you want to have this kind of message text, then you should define another list class and write a dedicated print function for it.  Then you have nice console messages AND usable numerical output.

File: R/representativity.R [hunk 14]
```diff
@@ -0,0 +1,102 @@
+#' Checks the representativity of the non-adult age groups in a mortAAR
+#' life table
+#'
+#' Weiss 1973 and Bocquet-Appel and Masset 1978 have devised indices
```
Comment (nevrome) [inline R/representativity.R]: Citation.

**Review by nevrome — CHANGES_REQUESTED** [review]
Very, very nice! 

The only really important change I suggest: Please design usable return values and not just console messages.

File: R/further_analytical_functions.R (hunk 1 above)
Comment (nmueller18) [reply R/further_analytical_functions.R]: Done!

File: R/further_analytical_functions.R (hunk 3 above)
Comment (nmueller18) [reply R/further_analytical_functions.R]: Done.

File: R/further_analytical_functions.R (hunk 2 above)
Comment (nmueller18) [reply R/further_analytical_functions.R:118]: Added comments.

File: R/further_analytical_functions.R (hunk 5 above)
Comment (nmueller18) [reply R/further_analytical_functions.R:146]: Done.

File: R/further_analytical_functions.R (hunk 10 above)
Comment (nmueller18) [reply R/further_analytical_functions.R]: Done.

File: R/lifetable_indices.R (hunk 11 above)
Comment (nmueller18) [reply R/lifetable_indices.R:17]: Okay, I have changed the code to make use of these already existing functions. The function itself is necessary, in my opinion, because it might not be self-evident for the user that she cannot use the mortaar_life_table_list as input. I am open for any suggestions, however.

File: R/further_analytical_functions.R (hunk 4 above)
Comment (nmueller18) [reply R/further_analytical_functions.R]: References added.

File: R/representativity.R (hunk 14 above)
Comment (nmueller18) [reply R/representativity.R]: References added.

File: R/lifetable_indices.R (hunk 12 above)
Comment (nmueller18) [reply R/lifetable_indices.R:51]: Yes and no. The indices are so important that I would like to make them visible for the user. But I have split the function to hide the mortality indices and to make the output and the referencing easier to read.

File: R/representativity.R (hunk 13 above)
Comment (nmueller18) [reply R/representativity.R]: Sounds sensible. BUT: How would such a list class and a dedicated print function look like?

File: R/further_analytical_functions.R (hunk 5 above)
Comment (nevrome) [reply R/further_analytical_functions.R:146]: Nice! But do we still need the first argument, the `mortaar_life_table_list`, in this case?

File: R/lifetable_indices.R (hunk 11 above)
Comment (nevrome) [reply R/lifetable_indices.R:17]: Ah - I understand. Hm.

File: R/representativity.R (hunk 13 above)
Comment (nevrome) [reply R/representativity.R]: If I understand that correctly, this function calculates a lot of different individual values. The output for example could be a list of (the new) class  `representativity_output`. 

```
foo <- list(
  index_of_children = 2, #round(indx$child_i$d5_9_d10_14, 2),
  index_of_juveniles = 3 #round(indx$juvenile_i$d5_14_d20plus,2)
) %>% `class<-`("representativity_output")

format.representativity_output <- function(x, ...) {
  return_value <- paste(x$index_of_children, "is a rainbow in the dark.")
  invisible(return_value)
}

print.representativity_output <- function(x, ...) cat(format(x, ...), "\n")

foo
```

Take a look at R/class_functions.R for an example right in this package. More info [here](http://adv-r.had.co.nz/S3.html).

File: R/further_analytical_functions.R (hunk 5 above)
Comment (nmueller18) [reply R/further_analytical_functions.R:146]: No, it is deleted in the last commit.

File: R/representativity.R (hunk 13 above)
Comment (nevrome) [reply R/representativity.R]: `representativity_output` is not a good class name btw.. I guess you can come up with something better.

Comment (nmueller18) [reply R/representativity.R]: Thanks for the hint! I will try it this way.

File: R/further_analytical_functions.R (hunk 6 above)
Comment (nmueller18) [reply R/further_analytical_functions.R:173]: Yes, you are right. I noticed it already during setting up the examples. I will look for something appropriate ...

File: R/further_analytical_functions.R (hunk 7 above)
Comment (nmueller18) [reply R/further_analytical_functions.R]: Hopefully, I have succeeded in changing every instance from index to name.

File: R/further_analytical_functions.R (hunk 8 above)
Comment (nmueller18) [reply R/further_analytical_functions.R:211]: It now returns a data.frame.

File: R/further_analytical_functions.R (hunk 9 above)
Comment (nmueller18) [reply R/further_analytical_functions.R]: Done

File: R/representativity.R (hunk 13 above)
Comment (nmueller18) [reply R/representativity.R]: After rethinking the matter, I opted against the dedicated class-option and sticked to a simple, but hopefully informative and clear data.frame. What do you think?

**Comment by nmueller18** [comment]
@nevrome: This was the last commit. Now all your requests should be answered.

**Comment by nevrome** [comment]
Good Job, @nmueller18! Maybe you want to write some basic tests for the new functions. It's not necessary, but good practice.

**Comment by nmueller18** [comment]
I have to confess that I am a bit lost here. I have read what Hadley wrote on the topic, and this certainly makes sense from an abstract point of view. But I caught myself tempted to simply copy the whole code of a function to test the result. This is certainly not intended, is it?

**Comment by nevrome** [comment]
Oh sorry - I forgot to answer to this question. 

There are several different aspects the tests for one function can/should cover. 

1. It's usual to check the interface. That includes the structure of the output (e.g. *is the output a list with the following variables* or *a data.frame with the folowing rows and columns*) and the ways a function deals with wrong input (eg. *the function should work if I enter a data.frame, but it should fail if I enter another data structure* or *if the input vector is too short this warning is shown*). An example for this approach can be found [here](https://github.com/ISAAKiel/mortAAR/blob/master/tests/testthat/test_analytical_functions_api.R). 

2. Another approach is to test if the function does the calculation in the right/intended way. You can test that by comparing the output to a static example dataset as it is done [here](https://github.com/ISAAKiel/mortAAR/blob/master/tests/testthat/test_analytical_functions_calculations.R). Or you can simply compare the output to static values. That could look like this: 

```
nitra_prep <- prep.life.table(
  nitra, group="sex", agebeg = "age_start", ageend = "age_end"
)
nitra_life <- life.table(nitra_prep)
sexrel <- lt.sexrelation(nitra_life$f, nitra_life$m)

test_that("lt.sexrelation calculates the correct results", {
  expect_equal(sexrel["MI", "value"], 66.7)
})
```

Btw. this test fails, because sexrel$value is a factor. That makes no sense to me. So I guess testing uncovered the first implementation problem here? Testing helps you to define the input and output of your functions very precisely. I wish it would be less tedious to write the testing code.
